mgr/dashboard: Adding RGW Bucket Notification Dashboard for Grafana

anushruti10116 opened the pull request:
This commit add the panel for bucket notification dashboard.

Signed-off-by: Anushruti Sharma <anushruti10116@gmail.com>


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

yuvalif commented:
> Thanks @cloudbehl for the review comments. Hi @yuvalif since Ankush has suggested to remove the `topic` variable let me know
> 
> * If I can update the query to
>   ` sum by (topic) (ceph_rgw_topic_persistent_topic_size{cluster=~"$cluster"})`
> * Also if I can fit both the panel in same row

1) IMO, summarizing over topics does not give lots of value. these topics are sent to external systems (e.g. kafka brokers).
issues in one of these systems does not necessarily mean issues in another. maybe, if we don't specify the topic we can display the summary by default?
3) would be great to show bytes and entries in the same graph. but since their scale is very different, can we show two scales in one graph? maybe right y-axis on the lest and the other y-axis on the right?

afreen23 commented:
jenkins test dashboard

yuvalif commented:
> Hi @cloudbehl please review the PR
> 
> * I have removed the topic variable and modified the query
how would it work if there are 1K topics?
do we have something similar with buckets?

> * Also I have created 2 panels side by side for simplification
so, do we have 1 panel for size in KB and one for number of entries?

> * Updated unit of Persistent topic size

anushruti10116 commented:
> > Hi @cloudbehl please review the PR
> > 
> > * I have removed the topic variable and modified the query
> >   how would it work if there are 1K topics?
> >   do we have something similar with buckets?
>    @yuvalif currently the metric is not exposing any label for bucket. If its possible to add a label for bucket in the metric then we can modify the query.
> > * Also I have created 2 panels side by side for simplification
> >   so, do we have 1 panel for size in KB and one for number of entries?
> yes we have two panels 
<img width="1410" height="341" alt="Screenshot 2025-08-07 at 6 44 11 PM" src="https://github.com/user-attachments/assets/cf886984-b02c-4de0-b376-41f8edb7128d" />

> > * Updated unit of Persistent topic size

anushruti10116 commented:
jenkins test make check

anushruti10116 commented:
jenkins test api

yuvalif commented:
> > > Hi @cloudbehl please review the PR
> > > 
> > > * I have removed the topic variable and modified the query
> > >   how would it work if there are 1K topics?
> > >   do we have something similar with buckets?
> > >   @yuvalif currently the metric is not exposing any label for bucket. If its possible to add a label for bucket in the metric then we can modify the query.
> > > * Also I have created 2 panels side by side for simplification
> > >   so, do we have 1 panel for size in KB and one for number of entries?
> > >   yes we have two panels
> 
> <img alt="Screenshot 2025-08-07 at 6 44 11 PM" width="1410" height="341" src="https://private-user-images.githubusercontent.com/33089887/475692465-cf886984-b02c-4de0-b376-41f8edb7128d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5MTIyNDgsIm5iZiI6MTc1NDkxMTk0OCwicGF0aCI6Ii8zMzA4OTg4Ny80NzU2OTI0NjUtY2Y4ODY5ODQtYjAyYy00ZGUwLWIzNzYtNDFmOGVkYjcxMjhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODExVDExMzIyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5YTliODRiYTc1ZmY2ZWMyY2Q3ODAxZGVlZWRhZjNmY2E2NjUxOWRlNzgwMjFlMThmZGJlNzNmZmI2M2ZmMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NRdaKIwI9eWihZVR9yAdcB7cHxdNCNT1_3h7P3OavxI">
> > > * Updated unit of Persistent topic size


* I think side by side is ok, but why is it simpler than one on top of the other?
* If you don't put them in one panel, you can keep the size panel as bytes of kilobytes - should not matter much.
only reason I suggested to use kilobytes, is use one panel, because the scale of number of entries and kilobytes is similar.
* topics and bucket are not fully related. my question was how will we present a graph with 1K topics?
I think that the only option is to have a searchable dropdown to select the topic.
the question on buckets was to try to understand how do we show bucket counters in graphs, where we have thousands of buckets?

anushruti10116 commented:
@cloudbehl as yuval suggested we should have drop down to select topic for that we can create a variable and use it in the query. Do you think we can put these panel as part of https://github.com/ceph/ceph/blob/main/monitoring/ceph-mixin/dashboards_out/rgw-s3-analytics.json where these panel will make more sense and we can easily add topic variable also

anushruti10116 commented:
jenkins test dashboard

anushruti10116 commented:
jenkins test make check arm64

anushruti10116 commented:
jenkins test api

cloudbehl reviewed (commented):
The current panels for persistent_topic_len (pending notifications count) and persistent_topic_size (pending data volume) are useful, but to make them more operator-friendly:

- Legends should be placed on the right side for better readability and to avoid clutter over the graph.
- Legends should display the last values so that users can instantly see how many notifications and how much data is currently undelivered without hovering over the graph.
- This way, at a glance, the legend itself becomes an actionable summary of undelivered notifications (count) and undelivered data (bytes/MB/GB) per topic.

anushruti10116 commented:
jenkins test make check arm64

anushruti10116 commented:
jenkins test dashboard cephadm

anushruti10116 commented:
jenkins test dashboard

yuvalif commented on monitoring/ceph-mixin/dashboards_out/radosgw-overview.json:
```diff
@@ -1185,6 +1185,215 @@
+         "thresholds": [ ],
+         "timeFrom": null,
+         "timeShift": null,
+         "title": "Persistent Topic Size",
```
in the second panel, shouldn't the headline be "Pending Notifications Size by Topic" ?

aaSharma14 commented:
jenkins test make check arm64

aaSharma14 commented:
Known failures for ceph dashboard tests


[should link user with account.RGW users page link user with account test should link user with account](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/20263/testReport/junit/(root)/should%20link%20user%20with%20account/RGW_users_page_link_user_with_account_test_should_link_user_with_account/)
["before all" hook for "should open notification sidebar".Notification page "before all" hook for "should open notification sidebar"](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/20263/testReport/junit/(root)/%22before%20all%22%20hook%20for%20%22should%20open%20notification%20sidebar%22/Notification_page__before_all__hook_for__should_open_notification_sidebar_/)
["after all" hook for "should open notification sidebar".Notification page "after all" hook for "should open notification sidebar"](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/20263/testReport/junit/(root)/%22after%20all%22%20hook%20for%20%22should%20open%20notification%20sidebar%22/Notification_page__after_all__hook_for__should_open_notification_sidebar_/)

aaSharma14 commented:
jenkins test dashboard cephadm

aaSharma14 commented:
jenkins test dashboard cephadm

afreen23 commented:
```
Using url [https://download.fedoraproject.org/pub/fedora/linux/releases/40/Cloud/x86_64/images/Fedora-Cloud-Base-Generic.x86_64-40-1.14.qcow2...](https://download.fedoraproject.org/pub/fedora/linux/releases/40/Cloud/x86_64/images/Fedora-Cloud-Base-Generic.x86_64-40-1.14.qcow2...%1B[0m)
Grabbing image fedora40...
Image fedora40 not Added because Unable to download indicated image
``` cephadm tests

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
[Test Result](https://jenkins.ceph.com/job/ceph-dashboard-cephadm-e2e/17909/testReport/) (1 failure / -11)
[should start a daemon.Services page should start a daemon](https://jenkins.ceph.com/job/ceph-dashboard-cephadm-e2e/17909/testReport/junit/(root)/should%20start%20a%20daemon/Services_page_should_start_a_daemon/)
	
Identified problems
Dashboard End-to-End Test Failure (Cypress)

1 of 11 failed (9%)

cephadm

afreen23 commented:
cephadm and dashboard are known failures. Pr can go ahead!

batrick commented:
Please squash these "Fix review comment" commits in the future.

@aaSharma14 please do not merge PRs in such a state in the future. Refer to our code submission guidelines: https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#separate-your-changes

aaSharma14 commented:
> Please squash these "Fix review comment" commits in the future.
> 
> @aaSharma14 please do not merge PRs in such a state in the future. Refer to our code submission guidelines: https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#separate-your-changes

@batrick I didn't notice the un-squashed before merging the PR, apologies for that. Will keep in mind going forward. Thanks
